traffic_ops/traffic_ops_golang/origin/origins.go

```diff
@@ -487,3 +487,294 @@ func deleteQuery() string {
+		if userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+			return
+		}
+	}
+
+	userErr, sysErr, errCode = checkTenancy(&origin.TenantID, &origin.DeliveryServiceID, inf.Tx, inf.User)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	res, err := tx.Exec("DELETE FROM origin WHERE id=$1", id)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+		return
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, fmt.Errorf("origin delete: getting rows affected: %w", err))
+		return
+	}
+	if rowsAffected == 0 {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, fmt.Errorf("no rows deleted for origin"), nil)
+		return
+	}
+	if rowsAffected != 1 {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, fmt.Errorf("origin delete: multiple rows affected"), nil)
+		return
+	}
+
+	alerts := tc.CreateAlerts(tc.SuccessLevel, "origin was deleted.")
+	api.WriteAlerts(w, r, http.StatusOK, alerts)
+	return
+}
+
+// readAndValidateJsonStruct reads json body and validates json fields.
+func readAndValidateJsonStruct(r *http.Request, tx *sqlx.Tx) (tc.OriginV5, int, error) {
+	var origin tc.OriginV5
+	if err := json.NewDecoder(r.Body).Decode(&origin); err != nil {
+		userErr := fmt.Errorf("error decoding POST request body into OriginV5 struct %w", err)
+		return origin, http.StatusBadRequest, userErr
+	}
+
+	noSpaces := validation.NewStringRule(tovalidate.NoSpaces, "cannot contain spaces")
+	validProtocol := validation.NewStringRule(tovalidate.IsOneOfStringICase("http", "https"), "must be http or https")
+	portErr := "must be a valid integer between 1 and 65535"
+
+	// validate JSON body
+	errs := tovalidate.ToErrors(validation.Errors{
+		"cachegroupId":      validation.Validate(origin.CachegroupID, validation.Min(1)),
+		"coordinateId":      validation.Validate(origin.CoordinateID, validation.Min(1)),
+		"deliveryServiceId": validation.Validate(origin.DeliveryServiceID, validation.NotNil),
+		"fqdn":              validation.Validate(origin.FQDN, validation.Required, is.DNSName),
+		"ip6Address":        validation.Validate(origin.IP6Address, validation.NilOrNotEmpty, is.IPv6),
+		"ipAddress":         validation.Validate(origin.IPAddress, validation.NilOrNotEmpty, is.IPv4),
+		"name":              validation.Validate(origin.Name, validation.Required, noSpaces),
+		"port":              validation.Validate(origin.Port, validation.NilOrNotEmpty.Error(portErr), validation.Min(1).Error(portErr), validation.Max(65535).Error(portErr)),
+		"profileId":         validation.Validate(origin.ProfileID, validation.Min(1)),
+		"protocol":          validation.Validate(origin.Protocol, validation.Required, validProtocol),
+		"tenantId":          validation.Validate(origin.TenantID, validation.Min(1)),
```

Comment:
There are some problems with this validation. `validation.Min` considers empty values to be valid, and empty values are determined for `int` types using reflection to determine if the value is zero. What that means is that `"tenantId": 0` is considered to be passing a filter of `validation.Min(1)`, even though its value is less than one. So when I don't provide a Tenant, this endpoint tells me that I'm not authorized on the non-existent Tenant, instead of telling me `tenantId` is required. You'll need to combine that validation with `validation.Required`.

Also, the `deliveryServiceId` is not actually being validated at all - you're giving it the `validation.NotNil` rule, but as a non-pointer property it'll never be `nil` even if I don't provide it. So you need to make it required instead.